### ISSUE TITLE
Exploring new models

### ISSUE THREAD

**Issue by deepfakes** [deepfakes/faceswap-model#1]
There are many ways to generate faces, if you find some interesting models, discuss it here

**Comment by deepfakes** [comment]
Currently looking at:
 - http://kvfrans.com/variational-autoencoders-explained/
 - http://int8.io/variational-autoencoder-in-tensorflow/

**Comment by Clorr** [comment]
A thought as it comes...

Model is trained with warped images, should we also train it with modified colors ? Maybe it would improve the final  rendering

**Comment by shaoanlu** [comment]
I've modified the original auto-encoder into a GAN.
https://github.com/shaoanlu/faceswap-GAN

btw, I tried random channel_shift on warped images and got worse results (input/output color inconsistency).

**Comment by deepfakes** [comment]
Wow ! Bravo for tackling that part !

I'm currently on Christmas duties and can't really get into this for now. You can see with @Ganonmaster what to do next. If he is ok with your proposal, we will happily add your contribution to the main repo.

Note that as touching models is a big change, maybe there will be more discussions than usual, but be sure that such contribution is precious and we will take it with the care it deserves

**Comment by Clorr** [comment]
@Ganonmaster did you see that ? What do you think about it ?

**Comment by Ganonmaster** [comment]
The results look very promising. Especially the improvement in the eyeball direction is astonishing. I do not have a any experience with GANs myself, so it is difficult for me to comment on an academic level. 😅 
However, looking through the jupyter notebooks, I think this code can, and should be incorporated. However, because it would result in a non-trivial increase training time, I would opt to do so in a way that does not conflict with the existing model, but still allow the user to test it out should they want to.

**Comment by Clorr** [comment]
Great then ! 

I'm thinking about a plugin architecture, already for alignment. It won't be very sophisticated but at least it will enable us to have different behaviors.

**Comment by shaoanlu** [comment]
There are 2 conceivable problems in my mind about merging it into the main repo:

1. Code I wrote in the training part is somehow tensowflow-ish that different from `model.fit(...)` or `model.train_on_batch(...)` as usual keras code looks like. 
2. The perceptual loss involves a VGGFace package that requires keras 2.1.1. It may cause compatibility problems, e.g. API changes (It did happen in my environment), as well as GPU memory error during training.

The first one is easy to solve since there are lots of keras GAN example on github. But the VGGFace one is more convolved. I can't come up with a good solution at the moment. Or maybe we just put the jupyter notebook into a `./demo` or `./examples` folder where people who have interest will look at, instead of merging my code into the main pipeline.

Anyway, there's no particular hurry, let the bullets fly. I believe there will soon be more people releasing their models/architectures.

**Comment by Clorr** [comment]
Someone posted this on Reddit:
https://www.reddit.com/r/deepfakes/comments/7n4w9p/comparison_of_loss_functions_model_decoder/

It adds some layer to the model. What's your opinion on this guys?
